src/Workspaces/Remote/ServiceHub/Shared/ClientDirectStream.cs

```diff
@@ -26,27 +24,30 @@ internal sealed class ClientDirectStream : Stream
 
         private static readonly TimeSpan s_connectRetryInterval = TimeSpan.FromMilliseconds(20);
 
-        private readonly string _name;
-        private readonly NamedPipeClientStream _pipe;
-        private readonly Stream _stream;
-
-        public ClientDirectStream(string name)
-        {
-            // this type exists so that consumer doesn't need to care about all these arguments/flags to get good performance
-            _name = name;
-            _pipe = new NamedPipeClientStream(serverName: ".", pipeName: name, direction: PipeDirection.Out);
-            _stream = new BufferedStream(_pipe, BufferSize);
-        }
-
-        public string Name => _name;
-
         public static async Task WriteDataAsync<TData>(string streamName, TData data, Func<ObjectWriter, TData, CancellationToken, Task> dataWriter, CancellationToken cancellationToken)
         {
             try
             {
-                using var stream = await CreateStreamAsync(streamName, cancellationToken).ConfigureAwait(false);
+                var pipe = new NamedPipeClientStream(serverName: ".", pipeName: streamName, PipeDirection.Out);
 
-                using (var objectWriter = new ObjectWriter(stream, cancellationToken))
+                bool success = false;
+                try
+                {
+                    await ConnectPipeAsync(pipe, cancellationToken).ConfigureAwait(false);
+                    success = true;
+                }
+                finally
+                {
+                    if (!success)
+                    {
+                        pipe.Dispose();
+                    }
+                }
+
+                // Transfer ownership of the pipe to BufferedStream, it will dispose it:
+                using var stream = new BufferedStream(pipe, BufferSize);
+
+                using (var objectWriter = new ObjectWriter(stream, leaveOpen: true, cancellationToken))
```

Comment:
can we comment more fully why this is the right thing to pass in here i'm not really understanding what's happening here with lifetimes/ownership.

Reply:
`leaveOpen` flag is standard BCL pattern when passing a Stream (or other disposable) into some writer. If `leaveOpen` is true then the writer doesn't dispose (own) the stream when it's disposed. Otherwise it does.